src/lib/ScannerCommand.ts

```diff
@@ -1,130 +1,64 @@
-		if (this.parsedFlags.rulename && typeof this.parsedFlags.rulename === 'string') {
-			filters.push(new RulenameFilter([this.parsedFlags.rulename]));
-		}
-
-		// Create a filter for any provided engines.
-		if (this.parsedFlags.engine && stringArrayTypeGuard(this.parsedFlags.engine) && this.parsedFlags.engine.length) {
-			filters.push(new EngineFilter(this.parsedFlags.engine));
-		}
-
-		return filters;
-	}
-
-	protected displayInfo(msg: string, verboseOnly: boolean): void {
-		if (!verboseOnly || this.parsedFlags.verbose) {
-			this.log(msg);
-		}
-	}
-
-	protected displayWarning(msg: string, verboseOnly: boolean): void {
-		if (!verboseOnly || this.parsedFlags.verbose) {
-			this.warn(msg);
-		}
-	}
-
-	protected displayError(msg: string): void {
-		this.error(msg);
-	}
+	protected abstract createAction(logger: Logger, display: Display) : Action;
 
-	protected startSpinner(msg: string, status="Please Wait"): void {
-		this.spinner.start(msg, status);
-	}
+	public async run(): Promise<AnyJson> {
+		const inputs: Inputs = (await this.parse(this.ctor)).flags;
+		const logger: Logger = await Logger.child(this.ctor.name);
+		const display: Display = new UxDisplay(this, this.spinner, inputs.verbose as boolean);
 
-	protected updateSpinner(msg: string): void {
-		this.spinner.status = msg;
-	}
+		display.displayWarning(getMessage(BundleName.Common, 'surveyRequestMessage'));
+		this.buildEventListeners(display);
 
-	/* eslint-disable-next-line @typescript-eslint/no-unused-vars */
-	protected waitOnSpinner(msg: string): void {
-		// msg variable is thrown away - please don't send anything here.
-		this.spinner.status += ' .';
+		const action: Action = this.createAction(logger, display);
+		await action.validateInputs(inputs);
+		return await action.run(inputs);
 	}
 
-	protected stopSpinner(msg: string): void {
-		this.spinner.stop(msg);
+	// TODO: Refactor away from events and instead inject the "Display" as a dependency into all of the classes that emit events
+	private buildEventListeners(display: Display): void {
+		uxEvents.on(EVENTS.INFO_ALWAYS, (msg: string) => display.displayInfo(msg));
+		uxEvents.on(EVENTS.INFO_VERBOSE, (msg: string) => display.displayVerboseInfo(msg));
+		uxEvents.on(EVENTS.WARNING_ALWAYS, (msg: string) => display.displayWarning(msg));
+		uxEvents.on(EVENTS.WARNING_VERBOSE, (msg: string) => display.displayVerboseWarning(msg));
+		uxEvents.on(EVENTS.START_SPINNER, (msg: string, status: string) => display.spinnerStart(msg, status));
+		uxEvents.on(EVENTS.UPDATE_SPINNER, (msg: string) => display.spinnerUpdate(msg));
+		uxEvents.on(EVENTS.WAIT_ON_SPINNER, (_msg: string) => display.spinnerWait()); // eslint-disable-line @typescript-eslint/no-unused-vars
+		uxEvents.on(EVENTS.STOP_SPINNER, (msg: string) => display.spinnerStop(msg));
 	}
+}
 
-	protected async init(): Promise<void> {
-		await super.init();
-		this.buildEventListeners();
-	}
+export interface Action {
+	validateInputs(inputs: Inputs): Promise<void>;
```

Comment:
Question: If there are actions whose `validateInputs` method is just an empty body, does it make sense for `validateInputs` to be a method defined by the interface? Might it be cleaner to expect the action's `run()` method to perform whatever validation needs to be performed?

Reply:
Fair question. I think that this is a fundamental responsibility in which the ScannerCommand should control the order of to ensure uniformity of flow. That is ScannerCommand forces validate to be called before run. This helps us when we test know that run is working with post-validated inputs and makes testing the two separate from one another. Also having a class have to think about validation is a good thing - even if they have no inputs. If we every have a case where there are no inputs to validate... then their validation can either be a no -op or we can consider making an Abstract class which makes the default of validate-inputs a no-op. Java has a nice "default" option on interfaces which I don't think typescript has. But in this case - all commands validate... so I think what I have here makes sense for now.